From 7b7ca843601235e066a3ce4d5f55548ee3c6d072 Mon Sep 17 00:00:00 2001
From: Michal Kubecek <mkubecek@suse.cz>
Date: Mon, 11 Dec 2017 08:17:41 +0100
Subject: [PATCH 16/19] netlink: add netlink handler for spause (-A)

Implement "ethtool -A <dev>" subcommand using netlink interface command
ETHTOOL_CMD_SET_PARAMS.

Signed-off-by: Michal Kubecek <mkubecek@suse.cz>
---
 ethtool.c |  3 ++-
 netlink.c | 28 ++++++++++++++++++++++++++++
 netlink.h |  1 +
 3 files changed, 31 insertions(+), 1 deletion(-)

diff --git a/ethtool.c b/ethtool.c
index 998417d7e8e9..6829f080be0c 100644
--- a/ethtool.c
+++ b/ethtool.c
@@ -4927,6 +4927,7 @@ int nl_geee(struct cmd_context *ctx) { return 0; }
 int nl_gfec(struct cmd_context *ctx) { return 0; }
 int nl_scoalesce(struct cmd_context *ctx) { return 0; }
 int nl_sring(struct cmd_context *ctx) { return 0; }
+int nl_spause(struct cmd_context *ctx) { return 0; }
 #endif
 
 static const struct option {
@@ -4952,7 +4953,7 @@ static const struct option {
 	  "		[ msglvl %d[/%d] | type on|off ... [--] ]\n" },
 	{ "-a|--show-pause", 1, do_gpause, nl_gpause,
 	  "Show pause options" },
-	{ "-A|--pause", 1, do_spause, NULL,
+	{ "-A|--pause", 1, do_spause, nl_spause,
 	  "Set pause options",
 	  "		[ autoneg on|off ]\n"
 	  "		[ rx on|off ]\n"
diff --git a/netlink.c b/netlink.c
index 055d913e40f2..e86bf980ff56 100644
--- a/netlink.c
+++ b/netlink.c
@@ -2224,3 +2224,31 @@ int nl_sring(struct cmd_context *ctx)
 	return nl_set_param(ctx, "-G", sring_params, ETHA_PARAMS_RING,
 			    ETHA_RING_MAX);
 }
+
+static const struct param_parser spause_params[] = {
+	{
+		.arg		= "autoneg",
+		.type		= ETHA_PAUSE_AUTONEG,
+		.handler	= nl_parse_bool,
+		.min_argc	= 1,
+	},
+	{
+		.arg		= "rx",
+		.type		= ETHA_PAUSE_RX,
+		.handler	= nl_parse_bool,
+		.min_argc	= 1,
+	},
+	{
+		.arg		= "tx",
+		.type		= ETHA_PAUSE_TX,
+		.handler	= nl_parse_bool,
+		.min_argc	= 1,
+	},
+	{}
+};
+
+int nl_spause(struct cmd_context *ctx)
+{
+	return nl_set_param(ctx, "-A", spause_params, ETHA_PARAMS_PAUSE,
+			    ETHA_PAUSE_MAX);
+}
diff --git a/netlink.h b/netlink.h
index 1e79086e49fe..a94cfce9a424 100644
--- a/netlink.h
+++ b/netlink.h
@@ -18,5 +18,6 @@ int nl_geee(struct cmd_context *ctx);
 int nl_gfec(struct cmd_context *ctx);
 int nl_scoalesce(struct cmd_context *ctx);
 int nl_sring(struct cmd_context *ctx);
+int nl_spause(struct cmd_context *ctx);
 
 #endif /* ETHTOOL_NETLINK_H__ */
-- 
2.18.0

